src/coreclr/gc/gc.cpp

```diff
@@ -51309,14 +51390,37 @@ bool CFinalize::SplitFinalizationData (CFinalize* other_fq)
         memmove (&other_fq->m_Array[otherIndex], &m_Array[thisIndex + thisNewSize], sizeof(other_fq->m_Array[0])*otherSize);
         other_fq->SegQueueLimit (i) = &other_fq->m_Array[otherIndex + otherSize];
 
-        // we delete the moved half from this queue
+        // slide the unmoved half to its new position in the queue
+        // (this will delete the moved half once copies and m_FillPointers updates are completed)
         memmove (segQueue, &m_Array[thisIndex], sizeof(m_Array[0])*thisNewSize);
         segQueue += thisNewSize;
         newFillPointers[i] = segQueue;
     }
 
+    segQueue = m_EndArray;
+    for (int i = MaxSeg; i > FreeListSeg; i++)
+    {
+        size_t thisIndex = SegQueue (i) - m_Array;
+        size_t thisLimit = SegQueueLimit (i) - m_Array;
+        size_t thisSize = thisLimit - thisIndex;
+
+        // we move half to the other queue
+        size_t otherSize = thisSize / 2;
+        size_t otherIndex = other_fq->SegQueueLimit (i) - other_fq->m_Array;
+        size_t thisNewSize = thisSize - otherSize;
+
+        memmove (&other_fq->m_Array[otherIndex], &m_Array[thisIndex + thisNewSize], sizeof(other_fq->m_Array[0])*otherSize);
+        other_fq->SegQueue (i) = &other_fq->m_Array[otherIndex];
+
+        // slide the unmoved half to its new position in the queue
+        // (this will delete the moved half once copies and m_FillPointers updates are completed)
+        segQueue -= thisNewSize;
+        memmove (segQueue, &m_Array[thisIndex], sizeof(m_Array[0])*thisNewSize);
+        newFillPointers[i - 1] = segQueue;
+    }
```

Comment:
similarly this one though note the subtle diffrences

Reply:
factored out SplitSegment.  This is pretty difficult because the copying left-to-right vs right-to-left changes so many of the pieces.  This allows the logic for the source locations and sizes to be written once (and not try to return 6 values from a helper).

Note that my original attempt at the right-to-left portion did not adjust otherIndex to be the beginning of the destination region.